### PR TITLE
[APS-67] Introduce facade for offender details

### PR DESCRIPTION
**Pull request by danlivings-dxw**
> See [APS-67 on the Approved Premises Jira board](https://dsdmoj.atlassian.net/browse/APS-67).

This PR introduces a facade for retrieving offender details, allowing us to abstract to some degree over the source of this information, simplifying the planned transition away from the Community API in favour of the AP Delius Context API.

This facade is implemented such that it can be configured through the Spring property `data-sources.offender-details` or the environment variable `DATA-SOURCES_OFFENDER-DETAILS`, which accepts values of either `community_api` or `ap_delius_context_api`. This allows for control over when and in what environments the transition to the AP Delius Context API occurs, and allows us to rollback the transition without reverting significant code changes should problems occur.

Currently, this facade only supports the Community API for individual offenders, and it replicates the existing behaviour. Future work will expand the facade to support listings, as well as providing an implementation backed by the AP Delius Context API.

### CODE REVIEW

File: src/main/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/datasource/OffenderDetailsDataSource.kt [hunk 1]
```diff
@@ -42,12 +43,20 @@ class ConfiguredOffenderDetailsDataSource(
...
+  private val communityApiClient: CommunityApiClient,
+) : OffenderDetailsDataSource {
   override val name: OffenderDetailsDataSourceName
     get() = OffenderDetailsDataSourceName.COMMUNITY_API
 
   override fun getOffenderDetailSummary(crn: String): ClientResult<OffenderDetailSummary> {
-    throw NotImplementedError("Getting details for individual offenders from the Community API is not currently supported")
+    var offenderResponse = communityApiClient.getOffenderDetailSummaryWithWait(crn)
+
+    if (offenderResponse is ClientResult.Failure.PreemptiveCacheTimeout) {
```
Comment (pezholio) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/datasource/OffenderDetailsDataSource.kt:58]: I'm hoping we'll be able to make the preemptive caching go away soon, but it makes sense to keep it in place for the Community API calls in the meantime

File: src/test/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/unit/datasource/CommunityApiOffenderDetailsDataSourceTest.kt [hunk 2]
```diff
@@ -46,41 +47,63 @@ class CommunityApiOffenderDetailsDataSourceTest {
...
+      val successBody = UserOffenderAccess(
+        userRestricted = false,
+        userExcluded = false,
+        restrictionMessage = null,
+      )
+
+      return allClientResults(successBody).intoArgumentStream()
+    }
+
+    private fun <T> allClientResults(successBody: T): List<ClientResult<T>> = listOf(
```
Comment (pezholio) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/unit/datasource/CommunityApiOffenderDetailsDataSourceTest.kt:89]: This is nice!

**Review by pezholio — APPROVED** [review]
This is great, thanks! Nice neat reversible solution - look forward to seeing how it progresses 👍

**Review by edavey — APPROVED** [review]
Thanks Dan, this is really easy to follow and looks like a good way to switch these endpoints. 

I do find the `@ParameterizedTest`s with `@MethodSource`s quite hard to parse but I guess my eyes will become accustomed!

**Review by tahb — APPROVED** [review]
Nice abstraction and well crafted!
